js/source/geojson_source.js

```diff
@@ -34,9 +38,30 @@ GeoJSONSource.prototype = util.inherit(Source, {
         return this;
     },
 
-    update() {
-        if (this._dirty) this._updateData();
-        if (this._loaded) this._updateTiles();
+    onAdd(map) {
+        this.map = map;
+    },
+
+    loaded() {
+        return this._loaded && this._pyramid.loaded();
+    },
+
+    update(transform) {
+        if (this._dirty) {
+            this._updateData();
+        }
+
+        if (this._loaded) {
+            this._pyramid.update(this.used, transform);
+        }
+    },
+
+    render(layers, painter) {
+        if (this._loaded) {
+            this._pyramid.renderedIDs().forEach((id) => {
```

Comment:
maybe lets avoid forEach in sensitive functions like render? Changing this to a simple loop may not make a big difference but small things like this add up.